src/FakeItEasy/ExpressionExtensions.cs

```diff
@@ -1,21 +1,109 @@
+        /// This is often going to be used in the context of finding out 'what object/value is a some (sub-)expression referring to?'
+        /// 
+        /// Key observation is that walking the expression tree and evaluating directly can be much faster for simple expressions like:
+        /// 
+        /// - constant expressions (null, 1, "easy", etc.)
+        /// 
+        /// - local variables referenced from lambda expressions, such as 'fake' in the A.CallTo line:
+        /// 
+        ///     {
+        ///         var fake = A.Fake&lt;Something&gt;();
+        ///         A.CallTo(() => fake.DoStuff());
+        ///     }
+        ///     
+        /// - Expressions that are simple member accesses (field gets, property gets) on an object or class, such as
+        ///     String.Empty
+        ///     A&lt;SomethingA&gt;.Ignored
+        ///     A&lt;SomethingA&gt;._
+        ///     myObj.someProperty
+        ///     
+        /// - Trivial method calls with no arguments (A.ToString(), B.GetType(), Factory.Create())
+        ///     
+        /// </notes>
         /// <param name="expression">The expression to be evaluated.</param>
         /// <returns>The value returned from the delegate compiled from the expression.</returns>
         public static object Evaluate(this Expression expression)
         {
+            return EvaluateOptimized(expression);
+        }
+        
+        private static object EvaluateOptimized(this Expression expression)
+        {
+            if (expression == null)
+            {
+                return null;
+            }
+
+            switch (expression.NodeType)
+            {
+                case ExpressionType.Constant:
+                    return ((ConstantExpression)expression).Value;
+
+                case ExpressionType.MemberAccess:
+                    var memberExpression = (MemberExpression)expression;
+
+                    var fieldInfo = memberExpression.Member as FieldInfo;
+                    if (fieldInfo != null)
+                    {
+                        return fieldInfo.GetValue(EvaluateOptimized(memberExpression.Expression));
+                    }
+
+                    var propertyInfo = memberExpression.Member as PropertyInfo;
+                    if (propertyInfo != null)
+                    {
+                        // index = null: this is always fine since it's a MemberAccess expression, not an IndexExpression
+                        return propertyInfo.GetValue(EvaluateOptimized(memberExpression.Expression), null);
+                    }
+
+                    break;
+
+                case ExpressionType.Call:
+                    var callExpression = (MethodCallExpression)expression;
+
+                    // for now, handling only very trivial call expressions with no arguments, like 'GetBlarg()'
+                    // because anything else might get complicated quickly (method overloads etc.)
+                    if (callExpression.Arguments == null)

```

Comment:
Interesting, I wonder how it knows? I wasn't sure from the type (ReadOnlyCollection<Argument>) that it would not return null instead of an empty collection.